Remove some of the block pattern caching

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Scale back on the caching of block patterns. Still do glob on every request. This way the code can easily detect the remove and adding of files in this directory. 

Trac ticket: https://core.trac.wordpress.org/ticket/59591
Trac ticket: https://core.trac.wordpress.org/ticket/59633

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Refactor.

spacedmonkey committed:
Reset.

felixarntz reviewed (changes requested):
Thanks @spacedmonkey, the approach certainly helps detecting new and deleted patterns, but it's not a proper solution for the root of the issue. When a pattern is updated, it will still be in the cache.

FWIW I disagree that any of these changes need to be made. This is a half baked solution. Either we should be able to rely on `WP_DEVELOPMENT_MODE`, or we should not (in which case _all_ of the block theme file caching in core would need to be reverted/removed).

Despite the above, what you're doing here partially addresses the concerns on the ticket, so I'm not strongly opposed to merging this as a small improvement. But we should have unit tests to make sure it doesn't introduce problems, especially this late in the cycle.

spacedmonkey commented:
>Despite the above, what you're doing here partially addresses the concerns on the ticket, so I'm not strongly opposed to merging this as a small improvement. But we should have unit tests to make sure it doesn't introduce problems, especially this late in the cycle.

@felixarntz You mind doing some performance profiling on this change? I believe one glob is better than 50 files exist for performance. Sadly I will not have time to work on this, as I am busy with other things.
